CAIPs/caip-294.md

```diff
@@ -0,0 +1,431 @@
+
+This provides the foundation for any Wallet Provider to interface with a Decentralized Application using a Blockchain Library which implements this standard.
+
+Different loading times can be affected by multiple factors, which makes it non-deterministic to publish and listen to messages from different sources within the browser.
+
+#### Discovery
+
+Both Wallet Providers and blockchain libraries must listen to incoming messages that might be published after their initialization. Additionally both Wallet Providers and blockchain libraries must publish a message to both announce themselves and their intent to connect, respectively.
+
+Here is the expected logic from the Blockchain Library:
+
+```typescript
+interface WalletMapEntry {
+  params: WalletAnnounceRequestParams;
+  eventName: string;
+}
+
+const wallets: Record<string, WalletMapEntry> = {}
+
+// Blockchain Library starts listening on init
+window.addEventListener("caip294:wallet_announce", (event) => {
+  // when an announce message was received then the library can index it by uuid
+  wallets[event.detail.params.uuid] = {
+    params: event.detail.params,
+    eventName: event.detail.params.uuid
+  }
+});
+
+// Blockchain Library publishes on init
+window.dispatchEvent(new CustomEvent(
+  "caip294:wallet_prompt",
+  {
+    detail: {
+      id: 1,
+      jsonrpc: "2.0"
+      method: "wallet_prompt",
+      params: {
+        // optionally the Blockchain Library can prompt wallets to announce matching only the chains
+        chains: []  // optional
+        //  if the Blockchain Library supports CAIP-275 then it can include a name
+        authName: "", // optional
+      },
+    }
+  }
+));
+```
+
+Here is the expected logic from the Wallet Provider:
+
+```typescript
+// Wallet Provider sets data on init
+const walletData = { ... }
+
+// Wallet Provider publishes on init
+window.dispatchEvent(new CustomEvent(
+  "caip294:wallet_announce",
+  {
+    detail: {
+      id: 1,
+      jsonrpc: "2.0"
+      method: "wallet_announce",
+      params: walletData,
+    }
+  }
+));
+
+
+
+// Wallet Providers starts listenning on init
+window.addEventListener("caip294:wallet_prompt", (event) => {
+  // when a prompt message was received then the wallet will announces again
+  window.dispatchEvent(new CustomEvent(
+    "caip294:wallet_announce",
+    {
+      detail: {
+        id: 1,
+        jsonrpc: "2.0"
+        method: "wallet_announce",
+        params: walletData,
+      }
+    }
+  ));
+});
+```
+
+#### Handshake
+
+After the wallet has been selected by the user then the Blockchain Library MUST publish a message to share its intent to establish a connection. This can be either done as a [CAIP-25][caip-25] request.
```

Comment:
We've been questioning this a bit with wallet-standard. Generally a wallet will allow a user to authorise some set of accounts for a dapp. It'll then make those available as accounts on the wallet object (or here the scopes object mentioned in caip-282). From the dapp's perspective, the user selecting that wallet (or the dapp selecting it for them) decides which accounts/features/chains/etc are available to it. But for the wallet, it only needs to care about a connection when it needs to change those scopes. That doesn't necessarily have to happen every time the user selects the wallet in an app.